pyflakes/test/test_code_segment.py

```diff
@@ -0,0 +1,139 @@
+                x = 0
+        '''
+        self.flakes(TestCodeSegments.parse_ast(code).body[0],
+                    m.UnusedVariable,
+                    scope=FunctionScope)
+
+    def test_class_segment(self):
+        code = '''
+        class Foo:
+            class Bar:
+                pass
+        '''
+        self.flakes(TestCodeSegments.parse_ast(code).body[0],
+                    scope=ClassScope)
+
+        code = '''
+        class Foo:
+            def bar():
+                x = 0
+        '''
+        self.flakes(TestCodeSegments.parse_ast(code).body[0],
+                    m.UnusedVariable,
+                    scope=ClassScope)
+
+    def test_scope_class(self):
+        code = '''
+        class Foo:
+            def bar(a, b=1, *d, **e):
+                pass
+        '''
+        checker = self.flakes(TestCodeSegments.parse_ast(code).body[0],
+                              scope=ClassScope)
+
+        scopes = checker.deadScopes
+        module_scopes = [
+            scope for scope in scopes if scope.__class__ is ModuleScope]
+        class_scopes = [
+            scope for scope in scopes if scope.__class__ is ClassScope]
+        function_scopes = [
+            scope for scope in scopes if scope.__class__ is FunctionScope]
+
+        # Ensure module scope is not present because we are analysing
+        # the inner contents of Foo
+        self.assertEqual(len(module_scopes), 0)
+        self.assertEqual(len(class_scopes), 1)
+        self.assertEqual(len(function_scopes), 1)
+
+        class_scope = class_scopes[0]
+        function_scope = function_scopes[0]
+
+        self.assertIsInstance(class_scope, ClassScope)
+        self.assertIsInstance(function_scope, FunctionScope)
+
+        self.assertIn('bar', class_scope)
+
+        self.assertIn('a', function_scope)
+        self.assertIn('b', function_scope)
+        self.assertIn('d', function_scope)
+        self.assertIn('e', function_scope)
+
+    def test_scope_function(self):
+        code = '''
+        def foo(a, b=1, *d, **e):
+            def bar(f, g=1, *h, **i):
+                pass
+        '''
+        checker = self.flakes(TestCodeSegments.parse_ast(code).body[0],
+                              scope=FunctionScope)
+
+        scopes = checker.deadScopes
+        module_scopes = [
+            scope for scope in scopes if scope.__class__ is ModuleScope]
+        function_scopes = [
+            scope for scope in scopes if scope.__class__ is FunctionScope]
+
+        # Ensure module scope is not present because we are analysing
+        # the inner contents of foo
+        self.assertEqual(len(module_scopes), 0)
+        self.assertEqual(len(function_scopes), 2)
+
+        function_scope_foo = function_scopes[1]
+        function_scope_bar = function_scopes[0]
+
+        self.assertIsInstance(function_scope_foo, FunctionScope)
+        self.assertIsInstance(function_scope_bar, FunctionScope)
+
+        self.assertIn('a', function_scope_foo)
```

Comment:
Need to assert which binding classes each of these things was.
(possibly assert all of them are the same, at least for `function_scope_bar`'s params

Reply:
Resolved